Fix Issue #422: Combat QoL update

Phrosfire opened the pull request:
# Pull Request

Pretty big change to combat for quality of life.

- Combat Arena is uncapped.
- Level differences are now taken into account for PvP streaks.  If someone attacks a player more than 10 levels under them and kills them, no changes will occur to the streak.  If the attacker loses, the streak is still affected.  If the attacker is more than 10 levels under the defender, streaks will still be affected. 
- +5% exp from PvP for each kill streak.  Example: A streak of 10 kills is a 50% PvP exp bonus.
- Max exp from PvP is 100.
- Justu exp has been increased to 50 when used in PvP.

Fix #422 

## License

By making this pull request, I confirm that I have the right to waive copyright and related rights to my contribution, and agree that all copyright and related rights in my contributions are waived, and I acknowledge that the Studie-Tech ApS organization has the copyright to use and modify my contribution for perpetuity.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
  - Arena fights are now accessible without daily restrictions.
  - The arena display now shows the daily fight count without referring to any training limits.
  - Increased the daily limit for battles in the arena to allow for a significantly higher number of fights per day.
  - Introduced a streak bonus for experience gain based on the user's PVP streak, enhancing the experience calculation.
  - Improved transfer level functionality with dynamic feedback on free transfers and reputation costs.

- **Bug Fixes**
  - Adjusted experience gain for "COMBAT" battle type to be capped and influenced by the user's PVP streak.
  - Enhanced data consistency by refreshing relevant user data after successful level transfers.

- **Style**
  - Updated text color for user roles and village information for better visibility in the UI.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MathiasGruber commented on app/src/libs/combat/util.ts:
```diff
@@ -618,15 +624,23 @@ export const calcBattleResult = (battle: CompleteBattle, userId: string) => {
+                        : (user.isAggressor || user.level - Math.max(...targets.map(t => t.level), 0) > -STREAK_LEVEL_DIFF 
+                            ? 0  // Reset streak for aggressors or if level difference is not extreme
+                            : user.pvpStreak))) // Keep streak if non-aggressor loses to someone STREAK_LEVEL_DIFF+ levels higher
+                : user.pvpStreak,
```
Maybe factor this out to a separate function that is easier to understand? Something like (if I read it right):

```typescript
const calculatePvpStreak = (
    battleType: string,
    user: { level: number; pvpStreak: number; isAggressor: boolean },
    targets: { level: number }[],
    didWin: boolean
): number => {
    if (battleType !== "COMBAT") {
        return user.pvpStreak;
    }

    const maxTargetLevel = Math.max(...targets.map(t => t.level), 0);
    const levelDifference = user.level - maxTargetLevel;

    if (user.isAggressor && levelDifference > STREAK_LEVEL_DIFF) {
        return user.pvpStreak;
    }

    if (didWin) {
        return user.pvpStreak + 1;
    }

    if (user.isAggressor || levelDifference > -STREAK_LEVEL_DIFF) {
        return 0;
    }

    return user.pvpStreak;
};
```

Phrosfire replied:
This change has been made and was tested to ensure it works.

MathiasGruber commented on app/src/libs/train.ts:
```diff
@@ -333,7 +333,7 @@ export const energyPerSecond = (speed: TrainingSpeed) => {
   switch (battleType) {
     case "COMBAT":
-      return experienceGain;
+      return 50;
```
Isn't this abuseable if you just keep fighting your own alt? the previous experience gain from such fights would be 0, but there's nothing dampening the gain when hard-coding? @theeneon ?

theeneon replied:
> Isn't this abuseable if you just keep fighting your own alt? the previous experience gain from such fights would be 0, but there's nothing dampening the gain when hard-coding? @theeneon ?

Overlooked that. I keep forgetting that persons will do something like this. Will have that introduced.

Phrosfire replied:
I adjusted this to use experienceGain and gave it the same modifier as pvp exp from killstreaks.  This way as players build streaks, their exp bonus will grow.  Since it uses experienceGain though, their jutsu exp from the fight will decrease if they kill the same person over and over.  The max jutsu exp gain is set to 50 as well.

Phrosfire commented:
I can work on the changes when I'm home from work

Phrosfire commented:
As part of the update, I also did a few other QoL changes.  The jutsu transfer screen now correctly updates after each transfer to show how many free transfer are left and specifies so when they will have to pay rep.  It wasn't updating properly unless the user manually refreshed.

I also adjusted the text color to black in a few village titles because they were hard to see.  I also made the coder and moderator text black in the title bar since they were hard to see against the role background color.  Kage was also adjusted to black text.

At present, normal fed usernames are nearly impossible to see on dark mode.  I adjusted the color of it to sky blue and it looks much better now.
![image](https://github.com/user-attachments/assets/6beaf8df-eec9-4b3c-a370-856d985994ce)

MathiasGruber reviewed (approved):
LGTM

MathiasGruber commented:
@Phrosfire can you force repush this? I have no idea what went wrong here - I tried to rebase locally on main, pushed, and somehow it just deleted all your commits. Hope you still have it locally 😬
